client/asset/btc/btc.go

```diff
@@ -1747,336 +1741,204 @@ func (btc *ExchangeWallet) LocktimeExpired(contract dex.Bytes) (bool, time.Time,
-		}
-		btc.log.Errorf("abandoning mempool redemption search for %d contracts because of %s",
-			contractsCount-totalFound-totalCanceled, reason)
+	if oldRedemption != nil {
+		oldRedemption.fail("Duplicate FindRedemption request received. Aborting the old request.")
 	}
 
-	mempoolTxs, err := btc.node.GetRawMempool()
-	if err != nil {
-		logAbandon(fmt.Sprintf("error retrieving transactions: %v", err))
-		return
-	}
+	go btc.tryRedemptionRequests(nil, []*findRedemptionReq{req})
 
-	for _, txHash := range mempoolTxs {
-		tx, err := btc.node.GetRawTransactionVerbose(txHash)
-		if err != nil {
-			logAbandon(fmt.Sprintf("getrawtransaction error for tx hash %v: %v", txHash, err))
-			return
-		}
-		found, canceled := btc.findRedemptionsInTx("mempool", tx, contractOutpoints)
-		totalFound += found
-		totalCanceled += canceled
-		if totalFound+totalCanceled == contractsCount {
-			break
+	var result *findRedemptionResult
+	select {
+	case result = <-req.resultChan:
+		if result == nil {
+			err = fmt.Errorf("unexpected nil result for redemption search for %s", outPt)
 		}
+	case <-ctx.Done():
+		err = fmt.Errorf("context cancelled during search for redemption for %s", outPt)
 	}
 
-	btc.log.Debugf("%d redemptions found, %d canceled out of %d contracts in mempool",
-		totalFound, totalCanceled, contractsCount)
+	// If this contract is still in the findRedemptionQueue, remove from the queue
+	// to prevent further redemption search attempts for this contract.
+	btc.findRedemptionMtx.Lock()
+	delete(btc.findRedemptionQueue, outPt)
+	btc.findRedemptionMtx.Unlock()
+
+	// result would be nil if ctx is canceled or the result channel is closed
+	// without data, which would happen if the redemption search is aborted when
+	// this ExchangeWallet is shut down.
+	if result != nil {
+		return result.redemptionCoinID, result.secret, result.err
+	}
+	return nil, nil, err
 }
 
-// findRedemptionsInBlockRange attempts to find spending info for the specified
-// contracts by searching every input of all txs in the provided block range.
-// If spending info is found for any contract, the contract is purged from the
-// findRedemptionQueue and the contract's secret (if successfully parsed) or any
-// error that occurs during parsing is returned to the redemption finder via the
-// registered result chan.
-// Also checks mempool for potential redemptions if spending info is not found
-// for any of these contracts in the specified block range.
-func (btc *ExchangeWallet) findRedemptionsInBlockRange(startBlock, endBlock *block, contractOutpoints []outPoint) {
-	contractsCount := len(contractOutpoints)
-	btc.log.Debugf("finding redemptions for %d contracts in blocks %d - %d",
-		contractsCount, startBlock.height, endBlock.height)
+// tryRedemptionRequests searches all mainchain blocks with height >= startBlock
+// for redemptions.
+func (btc *ExchangeWallet) tryRedemptionRequests(startBlock *chainhash.Hash, reqs []*findRedemptionReq) {
```

Comment:
On any of the returns in the function after error, wouldn't it be best to call `req.fail("epic fail")` in case someone is waiting and blocking for a value on the channel?